pkg/tbtc/node.go

```diff
@@ -205,3 +265,203 @@ func (n *node) joinDKGIfEligible(seed *big.Int, startBlockNumber uint64) {
+		// Additionally, we want to strongly extend the delay period
+		// periodically in order to give some additional time for nodes to
+		// recover and re-fill their internal TSS pre-parameters pools.
+		if drl.attemptCounter > 1 {
+			delayBlocks := drl.delayBlocks
+			if drl.attemptCounter%drl.delayBlocksBumpFrequency == 0 {
+				delayBlocks *= drl.delayBlocksBumpFactor
+			}
+
+			drl.attemptStartBlock = drl.attemptStartBlock +
+				dkg.ProtocolBlocks() +
+				delayBlocks
+		}
+
+		// Exclude all members controlled by the operators that were not
+		// qualified for the current attempt.
+		excludedMembers := make([]group.MemberIndex, 0)
+		attemptSkipped := false
+		for i, operator := range drl.selectedOperators {
+			if !qualifiedOperatorsSet[operator] {
+				memberIndex := group.MemberIndex(i + 1)
+				excludedMembers = append(excludedMembers, memberIndex)
+
+				// If the given member was not qualified for the given attempt,
+				// mark this attempt as skipped in order to skip the execution
+				// and set up the next attempt properly.
+				if memberIndex == drl.memberIndex {
+					attemptSkipped = true
+				}
+			}
+		}
+
+		var result *dkg.Result
+		var attemptErr error
+
+		if !attemptSkipped {
+			result, attemptErr = dkgAttemptFn(&dkgAttemptParams{
+				index:           drl.attemptCounter,
+				startBlock:      drl.attemptStartBlock,
+				excludedMembers: excludedMembers,
+			})
+			if attemptErr != nil {
+				var imErr *dkg.InactiveMembersError
+				if errors.As(attemptErr, &imErr) {
+					for _, memberIndex := range imErr.InactiveMembersIndexes {
+						operator := drl.selectedOperators[memberIndex-1]
+						drl.inactiveOperatorsSet[operator] = true
+					}
+				}
+			}
+		}
+
+		if attemptSkipped || attemptErr != nil {
+			var err error
+			qualifiedOperatorsSet, err = drl.qualifiedOperatorsSet()
+			if err != nil {
+				return nil, fmt.Errorf(
+					"cannot get qualified operators for attempt [%v]: [%w]",
+					drl.attemptCounter+1,
+					err,
+				)
+			}
+
+			continue
+		}
+
+		return result, nil
+	}
+}
+
+// qualifiedOperatorsSet returns a set of operators qualified to participate
+// in the given DKG attempt.
+func (drl *dkgRetryLoop) qualifiedOperatorsSet() (map[chain.Address]bool, error) {
+	// If this is one of the first attempts and random retries were not started
+	// yet, check if there are known inactive operators. If the group quorum
+	// can be maintained, just exclude the members controlled by the inactive
+	// operators from the qualified set.
+	if drl.attemptCounter <= 5 &&
+		drl.randomRetryCounter == 0 &&
+		len(drl.inactiveOperatorsSet) > 0 {
+		qualifiedOperators := make(chain.Addresses, 0)
+		for _, operator := range drl.selectedOperators {
+			if !drl.inactiveOperatorsSet[operator] {
+				qualifiedOperators = append(qualifiedOperators, operator)
+			}
+		}
+
+		if len(qualifiedOperators) >= drl.chainConfig.GroupQuorum {
+			return qualifiedOperators.Set(), nil
```

Comment:
Why are we building up a slice of operators and then converting it to a set, rather than building up that set directly?

ie:
```
qualifiedOperators := make(map[chain.Address]bool, len(drl.selectedOperators))
operatorCount := 0
for _, operator := range drl.selectedOperators {
	if !drl.inactiveOperatorsSet[operator] {
		qualifiedOperators[operator] = true
		operatorCount++
	}
}

if operatorCount >= drl.chainConfig.GroupQuorum {
	return qualifiedOperators, nil
}
```

Reply:
if we're going to immediately turn the list to a set, I can rewrite the retry API to return that directly and reduce complexity; I'm going out of the way to return a list at the moment

Reply:
Your version compares the distinct operators count to the `GroupQuorum`. However, we need to compare the number of members' seats. The `qualifiedOperators` slice is built from `drl.selectedOperators` which represents the group selection result which is about members' seats. I know this is a bit confusing as we are juggling between both representations of the group members (seats-oriented and operators-oriented) but I tried to follow a rule that every collection that has the `*Set` prefix is about distinct operators forming the group while everything else is about members seats (which are unfortunately represented using operators addresses as well).

Reply:
In mine, I keep track of the number of total seats via incrementing `operatorCount`, and then comparing `operatorCount >= drl.chainConfig.GroupQuorum` at the end, so while we might be overwriting the same `operator` key with true over and over, `operatorCount` keeps increasing. I think it should work!

> I tried to follow a rule that every collection that has the *Set suffix is about distinct operators forming the group while everything else is about members seats (which are unfortunately represented using operators addresses as well)

Yeah - totally understand the seats vs operators separation. I was making the point that this function `qualifiedOperatorsSet()` appears to be returning an operator Set, so we can build that Set up directly. I think it's more efficient computationally, and also easier to tell what the program is doing at a glance: make a set, and build it up vs make a list, build it up, convert it into a set.

Reply:
Oh yeah, I misread how `operatorCount` is constructed. However, I think keeping it as a slice is still ok as we are returning a slice from the retry API so this remains consistent. In the same time, I think that returning a slice form the retry API is also justified: https://github.com/keep-network/keep-core/pull/3169#discussion_r953889322

Reply:
Sounds good to me :100: